lib/phoenix/channel.ex

```diff
@@ -207,26 +203,40 @@ defmodule Phoenix.Channel do
   defmacro __using__(_) do
     quote do
       @behaviour unquote(__MODULE__)
+      @on_definition unquote(__MODULE__)
+      @before_compile unquote(__MODULE__)
       import unquote(__MODULE__)
       import Phoenix.Socket, only: [assign: 3]
+      Module.register_attribute(__MODULE__, :phoenix_handle_outs, accumulate: true)
 
       def handle_in(_event, _message, socket) do
         {:noreply, socket}
       end
 
-      def handle_out(event, message, socket) do
-        push(socket, event, message)
-        {:noreply, socket}
-      end
-
       def handle_info(_message, socket), do: {:noreply, socket}
 
       def terminate(_reason, _socket), do: :ok
 
-      defoverridable handle_info: 2, handle_out: 3, handle_in: 3, terminate: 2
+      defoverridable handle_info: 2, handle_in: 3, terminate: 2
+    end
+  end
+
+  defmacro __before_compile__(_) do
+    quote do
+      if @phoenix_handle_outs != [] do
+        def __fastlane__?(event) when event in @phoenix_handle_outs, do: false
+      end
+      def __fastlane__?(event), do: true
     end
   end
 
+  def __on_definition__(env, :def, :handle_out, [event, _payload, _socket], _, _)
+    when is_binary(event) do
+    Module.put_attribute(env.module, :phoenix_handle_outs, event)
```

Comment:
The problem with this solution is that someone can write:

    def handle_out(event, payload, socket) do
       # Add something to all messages
    end

And we are still going to fastlane. So the proper implementation would be, as soon as we find a handle_out clause that does not a binary event name, we need to assume we can no longer fastlane anything, at all.